docs/CONTRIBUTING.md

```diff
@@ -133,9 +135,76 @@ Please label issues with the following labels:
+band release any way, but this should be decided based on the severity on a case by case basis.
+
+### Versioning
+
+We are releasing multiple different things from this repository in one release, but 
+we don't want to use the same version for everything. Thus, in the following we explain
+the versioning story for the node, Westend & Rococo and the crates. The version associated
+to a particular release is taken from the node version.
+
+#### Node
+
+The versioning of the node is done 99% of the time by only incrementing the `minor` version. 
+The `major` version is only bumped for special releases and the `patch` can be used for an 
+out of band release that fixes some critical bug. The node version is not following SemVer. 
+This means that the version doesn't express if there are any breaking changes in the CLI 
+interface or similar. The node version is declared in the `NODE_VERSION` variable in 
+`polkadot/node/primitives/src/lib.rs`.
+
+#### Westend & Rococo
+
+For the test networks we only increment the `spec_version`. The spec version is also following
+the node version. So, `10020` is for example the node release `1.2.0`. This versioning has no
+further meaning and is only done to map from an on chain `spec_version` easily to the 
+release in this repository.
+
+#### Crate
+
+We want to follow SemVer as best as possible when it comes to crates versioning.
+Following SemVer isn't that easy, but there exists [a guide](https://doc.rust-lang.org/cargo/reference/semver.html) 
+in the Rust documentation that explains the small details on when to bump what. The bumping 
+should be done in between the releases in the PRs. However, it isn't required to 
+bump e.g. the `patch` version multiple times in between two releases. There
+exists a CI job that checks that the versions are not bumped multiple times to help 
+the developer. Another CI job is also checking for SemVer breaking changes. It is using
+[`cargo-semver-checks`](https://github.com/obi1kenobi/cargo-semver-checks). While 
+the tool isn't perfect, it should help to remind the developer of checking the SemVer 
+compatibility of its changes. In general there is not any guarantee to downstream 
+that there isn't a breaking in between and thus a `major` (or `minor` on `<1.0.0`) 
+bump. If possible, it should be prevented, but we also don't want to carry
+"dead code" with us for too long. As long as there are clear instructions on how 
+to integrate the breaking changes it should be fine to break things.
+
+So, the general approach is that developers are required to bump the versions in 
+all crates that they are changing. If there is a `major` (or `minor` on `<1.0.0`)
+crate version bump, it is important to also bump any crate that depends on this
+bumped version crate and re-export it. However, if the re-export is done in a
+`__private` module (that makes clear that it is internal api) the `major`/`minor`
+version bump doesn't ripple and only requires a `minor`/`patch` bump.
```

Comment:
I'm confused by this paragraph. Can you please explain the re-exporting? What is the `__private` module (is this just some example name?) Why do we need to bump a crate version if only the internal API changed, is it just because our tooling requires it?

Reply:
This is a public re-export for macros. This is a known pattern to make people on the outside aware that they should not use this module. However, from a pure SemVer perspective, this would be a breaking change.

Reply:
Oh I've never seen that pattern. I grepped and the `__private` looks like it's only used in substrate. But there is nothing substrate-specific about this doc or this "Crates" section, so it should be clarified. Also, I still don't understand this:

> the `major`/`minor` version bump doesn't ripple and only requires a `minor`/`patch` bump.

But, this paragraph should just be rewritten, and ideally should explain the "why". For sure it is very hard to understand, especially for new contributors.